### PR TITLE
Initial tenant cost API that is not wired up to any clients yet

### PR DESCRIPTION
**Pull request by oyving**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by oyving** [comment]
FYI - @olaaun

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 1]
```diff
@@ -354,6 +359,51 @@ private HttpResponse tenant(Tenant tenant, HttpRequest request) {
...
+        return tenantCost(tenant.get(), month.get(), request);
+    }
+
+    private Optional<LocalDate> tenantCostParseMonth(String monthString) {
+        var formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+
+        try {
+            return Optional.of(LocalDate.parse(monthString, formatter).withDayOfMonth(1));
+        } catch (DateTimeParseException ignored) {
+            return Optional.empty();
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: It's better to just rethrow this as an `IllegalArgumentExeption` and remove the explicit `month.isEmpty()` check, the `handle` method will then convert this into a bad request response.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 2]
```diff
@@ -354,6 +359,51 @@ private HttpResponse tenant(Tenant tenant, HttpRequest request) {
...
+        }
+
+        if (month.isEmpty()) {
+            return ErrorResponse.badRequest("Could not parse month '" + monthString + "'");
+        }
+
+        return tenantCost(tenant.get(), month.get(), request);
+    }
+
+    private Optional<LocalDate> tenantCostParseMonth(String monthString) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Looks like this is actually parsing a date string, not just month?

**Review by mpolden — CHANGES_REQUESTED** [review]

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 3]
```diff
@@ -198,6 +201,8 @@ private HttpResponse handleGET(Path path, HttpRequest request) {
         if (path.matches("/application/v4/user")) return authenticatedUser(request);
         if (path.matches("/application/v4/tenant")) return tenants(request);
         if (path.matches("/application/v4/tenant/{tenant}")) return tenant(path.get("tenant"), request);
+        if (path.matches("/application/v4/tenant/{tenant}/cost")) return tenantCost(path.get("tenant"), request);
+        if (path.matches("/application/v4/tenant/{tenant}/cost/{month}")) return tenantCost(path.get("tenant"), path.get("month"), request);
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Path parameter should be renamed.

**Review by mpolden — COMMENTED** [review]
There's a test failure:
```
[ERROR] Failures: 
[ERROR]   PathGroupTest.uniqueMatches:48 Paths '/application/v4/tenant/{tenant}/cost' and '/application/v4/tenant/{tenant}/cost/{*}' overlap.
```

**Review by mpolden — APPROVED** [review]

**Pull request merged**